### ISSUE TITLE
mwc-select does not work properly inside a mwc-dialog

### ISSUE THREAD

**Issue by holgerengels** [material-components/material-web#832]
**I'm submitting a:**

- [x] bug report
- [ ] feature request

**What OS are you using?**
Kubuntu

**What browser(s) is this bug affecting:**
Current Chrome

**Current behavior:**
mwc-select does not work inside a mwc-dialog. It's menu displays inside the dialog's content pane and makes it grow / show scrollbars.

**Expected behavior:**
the menu should display above the dialog

**Steps to reproduce:**
Open a dialog with a select with lots of items in it.

**Related code:**

```js
      <mwc-dialog id="dialog" title="Dialog">
        <form>
          <mwc-select>
            <mwc-list-item>01</mwc-list-item>
            <mwc-list-item>02</mwc-list-item>
            <mwc-list-item>03</mwc-list-item>
            <mwc-list-item>04</mwc-list-item>
            <mwc-list-item>05</mwc-list-item>
            <mwc-list-item>06</mwc-list-item>
            <mwc-list-item>07</mwc-list-item>
            <mwc-list-item>08</mwc-list-item>
            <mwc-list-item>09</mwc-list-item>
            <mwc-list-item>10</mwc-list-item>
            <mwc-list-item>11</mwc-list-item>
            <mwc-list-item>12</mwc-list-item>
            <mwc-list-item>13</mwc-list-item>
            <mwc-list-item>14</mwc-list-item>
            <mwc-list-item>15</mwc-list-item>
            <mwc-list-item>16</mwc-list-item>
            <mwc-list-item>17</mwc-list-item>
            <mwc-list-item>18</mwc-list-item>
            <mwc-list-item>19</mwc-list-item>
          </mwc-select>
        </form>
        <mwc-button slot="primaryAction" @click=${this._save}>Speichern</mwc-button>
      </mwc-dialog>
```

**Other information:**

**Comment by danielgek** [comment]
@aomarks I'm interested in fixing this one!

for what I could see, the dialog has `overflow: auto` in two places, this also happens when i try to put a select inside of a `mwc-top-app-bar-fixed` for the moment i got around it, but what would be your aproach, change the dropdown/menu behaviour or remove the overflows from the dialog and top bar ?

**Comment by govis** [comment]
Same problem with <code>mwc-menu</code> in a dialog. I think mwc-menu-surface might need to be absolute-positioned in a dialog by default.

**Comment by rwestlund** [comment]
I ran into this as well. Anyone have a temporary hack to make it work in the mean time?

**Comment by e111077** [comment]
moved up in priority. Actively investigating solutions now. Thanks for the input and activity on this issue

**Comment by e111077** [comment]
Update: this is still being worked on both upstream in MDC and in this repo. Unfortunately no PRs to track for now. This issue is still a good place to track

**Comment by c-dinkel** [comment]
We can use an `<mwc-menu fixed>` inside an `<mwc-dialog>` without encountering this issue, the menu renders correctly above the dialog that triggered it.

Seeing as `<mwc-select>` uses a `Menu` internally, could you maybe expose the `fixed` property through the `Select` API as a solution?

**Comment by e111077** [comment]
Unfortunately using fixed casues another host of behavioral problems. We are yet still working on a fix for this which is to implement an element hoisting system. It's just unfortunately stuck behind some upstream internal work

**Comment by c-dinkel** [comment]
Thanks for the update, while waiting for your internal work to complete we'll work around the issue by subclassing Select as you suggest and using *natural menu width* together with `fixed`.

**Comment by e111077** [comment]
Hello, a small update. On the canary channel we have implemented `fixedMenuPosition` on `mwc-select` which basically automatically does the workaround I described.

Not closing issue as we still want to eventually solve the menu hoisting issue

**Comment by yinonov** [comment]
in continuation of #2206 and relevance to here,
as it's still not public API,  
any chance on adding resize observer support? viewport change doesn't reposition / repaint the surface


https://user-images.githubusercontent.com/15140652/110529353-d4dba380-8121-11eb-9fb4-faea7b7c47d2.mov

**Comment by yinonov** [comment]
Thanks! not and an urgent matter. super cool if a thorough sweep is on the way

**Comment by c-dinkel** [comment]
@e111077 thanks for the updates. Am I understanding you correctly that in the canary workaround, a select inside a dialog opened on top of a document that has been scrolled will open its menu off-screen, while the scrim prevents the user from then scrolling up far enough to see it?

**Comment by e111077** [comment]
Hello, this should not be happening. In the workaround where you set fixedMenuPosition on the mwc-select, it should appear on the screen correctly when interacted with. If the Dom is scrolled or the window resized whilst the select's menu is open, then the menu will remain in its current position.

If the menu is closed, and the select opens it again, the menu should be correctly positioned again.

This is a big difference between fixed and absolute positioning.

The behavior you're describing seems to describe an absolute positioned menu

**Comment by e111077** [comment]
Periodic update that we haven't forgotten about this issue and still expect to have a solution down the line for this before 1.0.0

**Comment by lsatheeshkumar** [comment]
Is there an update on this please?

**Comment by e111077** [comment]
Internally at Google I have modified select, menu, and menu surface to make it easy to subclass and replace the menu surface with a native <dialog> element which has the power to render above everything.

I'll work on trying to externalize those changes so that y'all can do the same until the native browser popup proposal is adopted

**Comment by e111077** [comment]
Due to the feature freeze on MWC, I've only been able to do a code refactor on `mwc-select`, `mwc-menu`, and `mwc-menu-surface` just released as MWC `@0.27.0`.

With this refactor it should make it easier to subclass and replace `mwc-select`'s `mwc-menu`, and subsequent `mwc-menu-surface` to use a native `<dialog>` element with `HTMLDialogElement.showModal()` which taps into a browser's special ability to hoist dom to the top layer.

Here is an example implementation of this:

https://lit.dev/playground/#gist=f740bdff9ea4e62a64fae47bbdf5797e

Some caveats:

- when the menu is open, the rest of the DOM is inert (on mac voiceover, it seems to announce the same as before)
- when the menu is open, it will always scroll the menu
- scrolling the page horizontally while the menu is open will behave similar to `fixedMenuPosition` behavior

Due to the Material Web team's focus on M3 and the feature freeze on MWC, I'm going to close this issue as "wontfix" in favor of the dialog workaround linked above. In the future, when the `popup` proposal is available in browsers, this subclass method should still work and allow you to migrate to that which is the true answer to menu-surface's issues.

**Comment by Bob1k** [comment]
Just wrap it with <div style="position: fixed;z-index: 1;"> and it will work:
<mwc-dialog heading="Test dialog">
  <div style="height: 40px">
    <div style="position: fixed;z-index: 1;">
      <mwc-select>
        <mwc-list-item>item 1</mwc-list-item>
        <mwc-list-item>item 2</mwc-list-item>
        <mwc-list-item>item 3</mwc-list-item>
        <mwc-list-item>item 4</mwc-list-item>
        <mwc-list-item>item 5</mwc-list-item>
      </mwc-select>
    </div>
  </div>
</mwc-dialog>
